### ISSUE TITLE
Unexpected intersection type constraint error related to union, generic type and private field

### ISSUE THREAD

**Issue by vilic** [microsoft/TypeScript#38236]
**TypeScript Version:**  3.9.1-rc/4.0.0-dev.20200428

3.9.0-beta seems to be fine.

**Search Terms:** intersection,private,generic,constraint,union

**Code**

```ts
type Type = {foo: string} | {bar: number};

class SyncableObject {
  private foo: unknown;
}

interface SyncableRef<T extends ISyncableObject> {
}

interface ISyncableObject<T = object> extends SyncableObject {
}

type __ValueDescriptorType<
  T extends Type | undefined
> = T extends ISyncableObject
  ? SyncableRef<T>
  : T;
```

**Expected behavior:** No error

**Actual behavior:**

```
Type 'ISyncableObject<object> & T' does not satisfy the constraint 'ISyncableObject<object>'.
  Property 'foo' is missing in type 'ISyncableObject<object> & T' but required in type 'ISyncableObject<object>'.(2344)
```

**Playground Link:** [Playground](https://www.typescriptlang.org/play/index.html?ts=4.0.0-dev.20200429&ssl=1&ssc=1&pln=18&pc=1#code/C4TwDgpgBAKu0F4oG8BmB7dAuKBnYATgJYB2A5gL5QA+KARgIYE4kCuAtnRARQNwBQ-AMYAbBrlxQAyiBJCGdERADydAFYQhwFPyhQwxAG4Ng0DNiisSAaxLoA7iQEVBpUwVQMh0GXIVKAJQhUAB4YKAgAD1MSABNJAElfeUUVdU1gAD4dF343bk9vKCTZFKVVDS0wqCR0dK1sqJj46VL-NMrtZH5c0EgoAH0BgDUGEVYIABEIXCFiMGB0AjhIEN1YCOiIOMkV6ForWODSCFj+bKRwpu2Wkr9Uioz1gH5W+8DgsMz1nBgBIA)

**Related Issues:**

**Comment by DanielRosenwasser** [comment]
I'm guessing related to https://github.com/microsoft/TypeScript/pull/37762.

**Comment by ahejlsberg** [comment]
Actually, looks to be caused by #37537. We currently say that unions of object types never contain private or protected members. But that isn't quite right. A union of object types should contain any private member that is present in every constituent and originates in exactly the same declaration. This rarely matters, but in this particular case we end up structurally relating the union `ISyncableObject & object | ISyncableObject & string` to `ISyncableObject`, and that fails (but shouldn't) because of the private member.
